datafusion/core/src/physical_plan/coalesce_partitions.rs

```diff
@@ -199,6 +200,25 @@ impl Stream for MergeStream {
         cx: &mut std::task::Context<'_>,
     ) -> Poll<Option<Self::Item>> {
         let poll = self.input.poll_recv(cx);
+
+        // If the input stream is done, wait for all tasks to finish and return
+        // the failure if any.
+        if let Poll::Ready(None) = poll {
+            match Box::pin(self.tasks.join_next()).poll_unpin(cx) {
```

Comment:
I think this would help https://github.com/tokio-rs/tokio/pull/5721

Reply:
While I agree that it makes sense to add a `poll_join_next` method, you don't have to box here. You can use the `pin!` macro instead of boxing.